app/models/mhv_account.rb

```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+require 'mhv_ac/client'
+class MhvAccount < ActiveRecord::Base
+  include AASM
+
+  # Everything except ineligible accounts should be able to transition to :needs_terms_acceptance
+  ALL_STATES = %i(unknown needs_terms_acceptance ineligible registered upgraded register_failed upgrade_failed).freeze
+  after_initialize :setup
+
+  aasm(:account_state) do
+    state :unknown, initial: true
+    state :needs_terms_acceptance, :ineligible, :registered, :upgraded, :register_failed, :upgrade_failed
+
+    event :check_eligibility do
+      transitions from: ALL_STATES, to: :ineligible, unless: :mhv_account_eligible?
+      transitions from: ALL_STATES, to: :upgraded, if: :previously_upgraded?
+      transitions from: ALL_STATES, to: :registered, if: :previously_registered?
+      transitions from: ALL_STATES, to: :unknown
+    end
+
+    event :check_terms_acceptance do
+      transitions from: ALL_STATES - [:ineligible], to: :needs_terms_acceptance, unless: :terms_and_conditions_accepted?
+    end
+
+    event :register do
+      transitions from: [:unknown, :register_failed], to: :registered
+    end
+
+    event :upgrade do
+      transitions from: [:unknown, :registered, :upgrade_failed], to: :upgraded
+    end
+  end
+
+  def create_and_upgrade!
+    create_mhv_account! unless preexisting_account?
+    upgrade_mhv_account!
+  end
+
+  def terms_and_conditions_accepted?
+    terms_and_conditions_accepted.present?
+  end
+
+  def preexisting_account?
+    user&.mhv_correlation_id.present?
+  end
+
+  private
+
+  def terms_and_conditions_accepted
+    @terms_and_conditions_accepted ||=
+      TermsAndConditionsAcceptance.joins(:terms_and_conditions)
+                                  .includes(:terms_and_conditions)
+                                  .where(terms_and_conditions: { latest: true, name: 'mhv_account_terms' })
```

Comment:
I've been calling this `mhvac` in my code, but I could change it if you'd rather use this name.